### PR TITLE
Add native Shuffle integration

### LINKED ISSUE #15034: Shuffle integration
|Wazuh version|
|---|
| 4.4.0 |

## Description

Currently, Shuffle has developed a [custom integration](https://github.com/Shuffle/Shuffle/blob/launch/functions/extensions/wazuh/custom-shuffle.py) to forward alerts into their platform. The objective of this issue is to make it a native integration within Wazuh itself, like the ones for Slack or VirusTotal.

## Tasks

- [x] Review the code, test it and add it to Wazuh as a native integration.
- [x] Develop unit tests
- [x] Modify related Integratord files and unit tests to contemplate the new integration.
- [x] Verify the correct functioning of the integration when installing and updating Wazuh
- [x] Changelog.
- [x] Update the documentation.

### PR DESCRIPTION
**Pull request by fdalmaup**
|Related issue|
|---|
|#15034|

## Description

This PR closes #15034. It adds the Shuffle integration script as Wazuh native. Related installation files have been modified in order to only need the `ossec.conf` when in need to integrate with Shuffle. Also, unit tests have been carried out to check the behavior of the script.

## Configuration options
Inside `ossec.conf`, the following configuration is required to enable the integration
```
 <integration>
     <name>shuffle</name>
     <hook_url>http://<IP>:3001/api/v1/hooks/<HOOK_ID></hook_url>
     <level>3</level>
     <alert_format>json</alert_format>
 </integration>
```
## Unit Tests
```
============================= test session starts ==============================
platform linux -- Python 3.9.5, pytest-6.0.1, py-1.11.0, pluggy-0.13.1 -- /home/test_user/venv/unittest-env/bin/python3
cachedir: .pytest_cache
metadata: {'Python': '3.9.5', 'Platform': 'Linux-5.14.0-1052-oem-x86_64-with-glibc2.31', 'Packages': {'pytest': '6.0.1', 'py': '1.11.0', 'pluggy': '0.13.1'}, 'Plugins': {'cov': '3.0.0', 'trio': '0.7.0', 'metadata': '2.0.2', 'asyncio': '0.15.1', 'aiohttp': '0.3.0', 'html': '2.1.1'}}
rootdir: /home/test_user/git/wazuh
plugins: cov-3.0.0, trio-0.7.0, metadata-2.0.2, asyncio-0.15.1, aiohttp-0.3.0, html-2.1.1
collecting ... collected 18 items

integrations/tests/test_shuffle.py::test_main_bad_arguments_exit PASSED  [  5%]
integrations/tests/test_shuffle.py::test_main_exception PASSED           [ 11%]
integrations/tests/test_shuffle.py::test_main PASSED                     [ 16%]
integrations/tests/test_shuffle.py::test_process_args_exit[FileNotFoundError-3] PASSED [ 22%]
integrations/tests/test_shuffle.py::test_process_args_exit[side_effect1-4] PASSED [ 27%]
integrations/tests/test_shuffle.py::test_process_args PASSED             [ 33%]
integrations/tests/test_shuffle.py::test_process_args_not_sending_message PASSED [ 38%]
integrations/tests/test_shuffle.py::test_debug PASSED                    [ 44%]
integrations/tests/test_shuffle.py::test_generate_msg[87924-] PASSED     [ 50%]
integrations/tests/test_shuffle.py::test_generate_msg[rule-id-{"severity": 1, "pretext": "WAZUH Alert", "title": "alert description", "text": "full log.", "rule_id": "rule-id", "timestamp": "year-month-dayThours:minuts:seconds+0000", "id": "alert_id", "all_fields": {"timestamp": "year-month-dayThours:minuts:seconds+0000", "rule": {"level": 0, "description": "alert description", "id": "rule-id", "firedtimes": 1}, "id": "alert_id", "full_log": "full log.", "decoder": {"name": "decoder-name"}, "location": "wazuh-X"}}] PASSED [ 55%]
integrations/tests/test_shuffle.py::test_generate_msg_severity[3-1] PASSED [ 61%]
integrations/tests/test_shuffle.py::test_generate_msg_severity[5-2] PASSED [ 66%]
integrations/tests/test_shuffle.py::test_generate_msg_severity[7-2] PASSED [ 72%]
integrations/tests/test_shuffle.py::test_generate_msg_severity[8-3] PASSED [ 77%]
integrations/tests/test_shuffle.py::test_filter_msg[87924-False] PASSED  [ 83%]
integrations/tests/test_shuffle.py::test_filter_msg[rule-id-True] PASSED [ 88%]
integrations/tests/test_shuffle.py::test_send_msg_raise_exception PASSED [ 94%]
integrations/tests/test_shuffle.py::test_send_msg PASSED                 [100%]

============================== 18 passed in 0.08s ==============================
```

### CODE REVIEW

File: integrations/shuffle.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        f = open(log_file, "a")
+        f.write(msg)
+        f.close()
+
+
+# Skips container kills to stop self-recursion
+def filter_msg(alert) -> bool:
+    # These are things that recursively happen because Shuffle starts Docker containers
+    skip_rule_ids = ["87924", "87900", "87901", "87902", "87903", "87904", "86001", "86002", "86003", "87932", "80710", "87929",
+            "87928", "5710"]
```
Comment (CarlosRS9) [inline integrations/shuffle.py:97]: Move this to the `Global vars` section as a constant

File: integrations/shuffle.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    return not alert["rule"]["id"] in skip_rule_ids
+
+    # try:
+    #    if "docker" in alert["rule"]["description"].lower() and "
+    # msg['text'] = alert.get('full_log')
+    # except:
+    #    pass
+    # msg['title'] = alert['rule']['description'] if 'description' in alert['rule'] else "N/A"
```
Comment (CarlosRS9) [inline integrations/shuffle.py:105]: Why is this here?

File: integrations/shuffle.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        if len(sys.argv) >= 4:
+            msg = '{0} {1} {2} {3} {4}'.format(
+                now,
+                sys.argv[1],
+                sys.argv[2],
+                sys.argv[3],
+                sys.argv[4] if len(sys.argv) > 4 else '',
+            )
+            # debug_enabled = (len(sys.argv) > 4 and sys.argv[4] == 'debug')
+            debug_enabled = True
```
Comment (CarlosRS9) [inline integrations/shuffle.py:154]: Why is this always enabled?
```suggestion
            debug_enabled = (len(sys.argv) > 4 and sys.argv[4] == 'debug')
```

**Review by CarlosRS9 — CHANGES_REQUESTED** [review]

File: integrations/tests/test_shuffle.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+import pytest
+import shuffle
+import random
+
+sys.path.append(os.path.join(os.path.dirname(
+    os.path.realpath(__file__)), '..', '..'))
+
+filter_values = ["87924", "87900", "87901", "87902", "87903", "87904", "86001",
+                 "86002", "86003", "87932", "80710", "87929", "87928",
+                 "5710"]
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:19]: Use the constant from the shuffle.py module instead

File: integrations/tests/test_shuffle.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    Test that the expected message is generated when json_alert received.
+
+    Parameters
+    ----------
+    json_alert : json
+        json data that simulates the values that could have
+        been obtained from alert_file_location.
+
+    msg : str
+        message that should be retourned by the generate_msg function
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:49]:
```suggestion
        message that should be returned by the generate_msg function
```

File: integrations/tests/test_shuffle.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                                              '')])
+def test_generate_msg(json_alert, msg):
+    """
+    Test that the expected message is generated when json_alert received.
+
+    Parameters
+    ----------
+    json_alert : json
+        json data that simulates the values that could have
+        been obtained from alert_file_location.
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:46]:
```suggestion
        json data that simulates the values that could have been obtained from alert_file_location.
```

File: integrations/tests/test_shuffle.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                                             ({'timestamp': 'year-month-dayThours:minuts:seconds+0000',
+                                               'rule': {'level': 0, 'description': 'alert description',
+                                                        'id': random.choice(filter_values),
+                                                        'firedtimes': 1},
+                                               'id': 'alert_id',
+                                               'full_log': 'full log.', 'decoder': {'name': 'decoder-name'},
+                                               'location': 'wazuh-X'},
+                                              '')])
+def test_generate_msg(json_alert, msg):
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:38]: Do not hardcode events in the parametrize. The test cases you are using are identical except for the `rule-id`. Use variables for both the alert and the expected message and provide only the rule_id.

Take this as an example on how it could look like:

```suggestion
@pytest.mark.parametrize('alert, expected_msg, rule_id', [
    (alert_template, msg_template, 1),
    (alert_template, "", shuffle.SKIP_RULE_IDS[0])
])
def test_generate_msg(alert, expected_msg, rule_id):
```

File: integrations/tests/test_shuffle.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                                                         'location': 'wazuh-X'})])
+def test_filtered_msg(json_alert):
+    """
+    Test that the alerts with certain rule ids are filtered.
+
+    Parameters
+    ----------
+    json_alert : json
+        json data that simulates the values that could have
+        been obtained from alert_file_location.
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:69]:
```suggestion
        json data that simulates the values that could have been obtained from alert_file_location.
```

File: integrations/tests/test_shuffle.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+
+@pytest.mark.parametrize('json_alert', [({'timestamp': 'year-month-dayThours:minuts:seconds+0000',
+                                                         'rule': {'level': 0, 'description': 'alert description',
+                                                                  'id': 'rule-id',
+                                                                  'firedtimes': 1},
+                                                         'id': 'alert_id',
+                                                         'full_log': 'full log.', 'decoder': {'name': 'decoder-name'},
+                                                         'location': 'wazuh-X'})])
+def test_filtered_msg(json_alert):
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:61]: Apply the same changes as with the other test

File: integrations/tests/test_shuffle.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    Test that the alerts with certain rule ids are filtered.
+
+    Parameters
+    ----------
+    json_alert : json
+        json data that simulates the values that could have
+        been obtained from alert_file_location.
+
+    """
+    json_alert['rule']['id'] = random.choice(filter_values)
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:72]: we should not choose a random value. This could cause the test to work on some runs and fail on others. Consider always using the same id or traversing the entire list.

File: integrations/tests/test_shuffle.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+
+@pytest.mark.parametrize('json_alert', [({'timestamp': 'year-month-dayThours:minuts:seconds+0000',
+                                          'rule': {'level': 0, 'description': 'alert description',
+                                                   'id': 'rule-id',
+                                                   'firedtimes': 1},
+                                          'id': 'alert_id',
+                                          'full_log': 'full log.', 'decoder': {'name': 'decoder-name'},
+                                          'location': 'wazuh-X'})])
+def test_not_filtered_msg(json_alert):
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:84]: Same as the other tests

File: integrations/tests/test_shuffle.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    """
+    Test that the alerts that not contain certain rule ids are not filtered.
+
+    Parameters
+    ----------
+    json_alert : json
+        json data that simulates the values that could have
+        been obtained from alert_file_location.
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:92]:
```suggestion
        json data that simulates the values that could have been obtained from alert_file_location.
```

**Review by CarlosRS9 — CHANGES_REQUESTED** [review]

**Comment by CarlosRS9** [comment]
In addition to the requested changes you are also missing some tests cases for error exception handling. Also, please make sure you use the coverage option to ensure the tests have a converage of 100% if possible. You can use something like `--cov . --cov-report term-missing`.

File: integrations/shuffle.py (hunk 2 above)
Comment (fdalmaup) [reply integrations/shuffle.py:105]: Those comments were in the original script from [the shuffle repo](https://github.com/Shuffle/Shuffle/blob/launch/functions/extensions/wazuh/custom-shuffle.py). They could be part of an alternative behavior of the script but I will remove the section because it wasn't tested and does not contribute to the clarity of the code

File: integrations/shuffle.py (hunk 3 above)
Comment (fdalmaup) [reply integrations/shuffle.py:154]: I didn't verify the `debug_enabled` behavior that was in the comment. The suggested change will be tested

File: integrations/tests/test_shuffle.py [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+    Parameters
+    ----------
+    args: list[str]
+       list of the arguments passed to the main function
+
+    """
+    with patch("builtins.open", mock_open()), \
+            pytest.raises(SystemExit) as pytest_wrapped_e:
+        shuffle.main(args[0:2])
+    assert pytest_wrapped_e.type == SystemExit
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:47]: There is no need to perform this check . It is guaranteed that the exception will be a `SystemExit` one otherwise `pytest.raises` will fail and this assert sentence will not be executed.

File: integrations/tests/test_shuffle.py [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+    Parameters
+    ----------
+    args: list[str]
+       list of the arguments passed to the process_args function
+
+    """
+
+    with patch('shuffle.process_args') as process, \
+            patch("builtins.open", mock_open()), \
+            pytest.raises(Exception) as pytest_wrapped_e:
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:65]: No need to add `as pytest_wrapped_e` if you are not going to use it later.

File: integrations/tests/test_shuffle.py [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+def test_process_args_alert_file_exit(args):
+    """
+    Test that process_args function exits when alert file is not found
+
+    Parameters
+    ----------
+    args: list[str]
+       list of the arguments passed to the process_args function
+
+    -------
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:101]:
```suggestion
```

File: integrations/tests/test_shuffle.py [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+
+
+    """
+    log_file = tmpdir.join('test.log')
+    with patch('shuffle.debug_enabled', return_value=debug_enabled), \
+            patch('shuffle.LOG_FILE', str(log_file)), \
+            patch("builtins.open", mock_open()), \
+            pytest.raises(SystemExit) as pytest_wrapped_e:
+        shuffle.process_args(args)
+    assert pytest_wrapped_e.type == SystemExit
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:155]:
```suggestion
```
Same as before.

File: integrations/tests/test_shuffle.py [hunk 17]
```diff
@@ -0,0 +1,307 @@
...
+
+    alert: json
+       template alert read from the alert file
+
+
+    """
+    with patch("builtins.open", mock_open()), \
+            patch('json.load', return_value=alert), \
+            patch('requests.post', return_value=requests.Response):
+        shuffle.process_args(args)
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:128]: To test that the function is working as expected you must ensure that both `generate_msg` and `send_msg` functions are being invoked with the expected parameters.

File: integrations/tests/test_shuffle.py [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+    args: list[str]
+       list of the arguments passed to the process_args function
+
+    alert: json
+       template alert read from the alert file
+    """
+    with patch("builtins.open", mock_open()), \
+            patch('json.load', return_value=alert), \
+            patch('shuffle.generate_msg', return_value=''):
+        shuffle.process_args(args)
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:175]: You should ensure that `send_msg` function is not being called.

**Review by CarlosRS9 — CHANGES_REQUESTED** [review]

**Comment by CarlosRS9** [comment]
You are missing one test for the `send_msg` function to ensure it is using the `request` functionality as intended.

Also, please review the docstrings to ensure it follows our standard.

File: integrations/tests/test_shuffle.py [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+        process.side_effect = Exception
+        shuffle.main(sys_args_template)
+
+
+def test_main_correct_execution():
+    """Test the correct execution of the main function."""
+    with patch("builtins.open", mock_open()), \
+            patch('json.load', return_value=alert_template), \
+            patch('requests.post', return_value=requests.Response):
+        shuffle.main(sys_args_template)
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:59]: You are just invoking the main function and nothing else. Generally UTs have at least one assert or check.
In this case I would check that you are calling `process_args` with the expected arguments.

In addition, I would call this simply `test_main`

File: integrations/tests/test_shuffle.py [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+
+
+def test_process_args_alert_file_exit():
+    """Test that process_args function exits when alert file is not found."""
+    with patch("builtins.open") as open_mock, \
+            pytest.raises(SystemExit) as pytest_wrapped_e:
+        open_mock.side_effect = FileNotFoundError
+        shuffle.process_args(sys_args_template)
+    assert pytest_wrapped_e.value.code == 3
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:68]: You should combine this test with the `test_process_args_json_exit`. They are doing the same thing. You just need to use a parametrize to specify the side_effect and the expected error code. Try to use test cases for scenarios like those.

File: integrations/tests/test_shuffle.py [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+    ----------
+    tmpdir: py.path.local
+        Path to a temporary directory generated for the tests logs.
+    """
+    log_file = tmpdir.join('test.log')
+    with patch('shuffle.debug_enabled', return_value=True), \
+            patch('shuffle.LOG_FILE', str(log_file)):
+        shuffle.debug(msg_template)
+
+    assert log_file.read() == f"{shuffle.now}: {msg_template}\n"
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:126]: Never make use of temporary files in a unit test unless it is absolutely necessary. You can achieve the same thing by making that the open file returns a mocked object and checking `your_mock_object.write.assert_called_with(...)` instead of reading from a real log file. For example, you could use `MagickMock`.

File: integrations/tests/test_shuffle.py [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+    """
+    log_file = tmpdir.join('test.log')
+    with patch('shuffle.debug_enabled', return_value=True), \
+            patch('shuffle.LOG_FILE', str(log_file)), \
+            patch("builtins.open", mock_open()), \
+            patch('json.load') as json_load, \
+            pytest.raises(SystemExit) as pytest_wrapped_e:
+        json_load.side_effect = json.decoder.JSONDecodeError("Expecting value", "", 0)
+        shuffle.process_args(sys_args_template)
+    assert pytest_wrapped_e.value.code == 4
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:100]: Don't use temporary directories nor temporary files. This can be achieved using mocks instead.

File: integrations/tests/test_shuffle.py [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+    """Test that the alerts with certain rule ids are filtered."""
+    for skip_id in shuffle.SKIP_RULE_IDS:
+        alert_template['rule']['id'] = skip_id
+        assert not shuffle.filter_msg(alert_template)
+
+
+def test_not_filtered_msg():
+    """Test that the alerts that not contain certain rule ids are not filtered."""
+    alert_template['rule']['id'] = 'rule-id'
+    assert shuffle.filter_msg(alert_template)
```
Comment (CarlosRS9) [inline integrations/tests/test_shuffle.py:177]: Consider merging these two into a single one using different test cases.

**Review by CarlosRS9 — CHANGES_REQUESTED** [review]

**Review by CarlosRS9 — CHANGES_REQUESTED** [review]

**Review by CarlosRS9 — APPROVED** [review]

**Review by davidjiglesias — APPROVED** [review]
LGTM

**Review by jmv74211 — APPROVED** [review]
## QA review

- **Type**: Manual testing.
- **Status**: Approved  🟢 
- **Testing issue**: https://github.com/wazuh/wazuh-qa/issues/3398
- **Comments**:  https://github.com/wazuh/wazuh-qa/issues/3398#issuecomment-1290251173
